### PR TITLE
adding support for running tests on multiple  k8s versions

### PR DESCRIPTION
**Pull request by dangula**
Support running multiple k8s versions

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -20,7 +20,8 @@ pipeline {
                 sh 'build/run make -j\$(nproc) test'
             }
         }
-        stage('Integration Tests') {
+
+        stage('Smoke Test - k8s 1.6'){
```
Comment (bassam) [inline Jenkinsfile]: is this needed? let's just run the whole matrix.

File: Jenkinsfile [hunk 2]
```diff
@@ -29,6 +30,77 @@ pipeline {
                 sh 'build/run make -j\$(nproc) test-integration'
             }
         }
+
+        stage('Integration Tests SetUp') {
```
Comment (bassam) [inline Jenkinsfile]: I think you should combine the "setup" stage with the "run" stage

File: Jenkinsfile [hunk 3]
```diff
@@ -29,6 +30,77 @@ pipeline {
...
+            steps {
+                sh 'tests/scripts/saveImages.sh > BUILD_NUMBER'
```
Comment (bassam) [inline Jenkinsfile]: why is the BUILD_NUMBER needed? can you not tag the image to something stable in saveImages.sh?

File: Jenkinsfile [hunk 4]
```diff
@@ -29,6 +30,77 @@ pipeline {
...
+                    BUILD_NUM = readFile('BUILD_NUMBER').trim()
+                }
+                echo "Saved images from build ${BUILD_NUM}"
+                stash name: 'repo',excludes: '_output/,vendor/,.cache/,.work/'
+            }
+        }
+
+        stage('Integration Tests - parallel runs') {
+            environment {
+                KUBE_VERSION = "v1.6"
```
Comment (bassam) [inline Jenkinsfile]: why is this needed? its explicitly set below

File: Jenkinsfile [hunk 5]
```diff
@@ -29,6 +30,77 @@ pipeline {
...
+            }
+            steps {
+                parallel(
+                    "GCE_K8s_v1.7" : {
```
Comment (bassam) [inline Jenkinsfile]: can this be run in a loop? there must be a better way to capture a "matrix" in jenkins.

File: Jenkinsfile [hunk 6]
```diff
@@ -29,6 +30,77 @@ pipeline {
...
+                        node("gce_vm"){
+                            script {
+                                try{
+                                    withEnv(["KUBE_VERSION=v1.7"]){
+                                        deleteDir()
+                                        unstash 'repo'
+                                        sh "docker load -i rookamd64.tar"
```
Comment (bassam) [inline Jenkinsfile]: I think you need to add a `loadimages.sh` or better yet make it `test-images.sh save|load`. that way the naming is not split between script and jenkins and its more maintainable.

File: Jenkinsfile [hunk 7]
```diff
@@ -29,6 +30,77 @@ pipeline {
...
+                                        sh "docker load -i toolboxamd64.tar"
+                                        sh 'build/run make -j\$(nproc) build.common'
```
Comment (bassam) [inline Jenkinsfile]: why do you need build.common? can it be a makefile pre-req to test-integration?

File: Jenkinsfile [hunk 8]
```diff
@@ -29,6 +30,77 @@ pipeline {
...
+                                        sh "tests/scripts/kubeadm-dind.sh up ${BUILD_NUM}"
```
Comment (bassam) [inline Jenkinsfile]: why is `BUILD_NUM` needed here?

File: Jenkinsfile [hunk 9]
```diff
@@ -29,6 +30,77 @@ pipeline {
...
+                                        sh 'build/run make -j\$(nproc) test-integration'
+                                    }
+                                }
+                                finally{
+                                    archive '_output/tests/*'
```
Comment (bassam) [inline Jenkinsfile]: given that our tests make changes at the kernel level (rbd|cephfs) I don't think we can rely on the host machine not getting tainted after a test (especially a bad test or code). Can these agents that run integrations tests be restarted at the VM level between test runs?

File: tests/baseSetup.go [hunk 10]
```diff
@@ -36,6 +37,9 @@ var (
 //One init function per package - initializes Rook infra and installs rook(if needed based on flags)
 func init() {
 	Env = objects.NewManifest()
+	if kv := os.Getenv("KUBE_VERSION"); kv != "" {
```
Comment (bassam) [inline tests/baseSetup.go]: this should be a flag that comes from the environment. lots of examples to follow from cmd/rook/main.go

File: tests/scripts/kubeadm-dind.sh [hunk 11]
```diff
@@ -40,8 +40,9 @@ export CNI_PLUGIN="${CNI_PLUGIN:-bridge}"
 case "${1:-}" in
   up)
     ${scriptdir}/dind-cluster.sh reup
-    copy_image_to_cluster ${BUILD_REGISTRY}/rook-amd64 rook/rook:master
-    copy_image_to_cluster ${BUILD_REGISTRY}/toolbox-amd64 rook/toolbox:master
+    BUILD_NUM=${2:-${BUILD_REGISTRY}}
+    copy_image_to_cluster ${BUILD_NUM}/rook-amd64:latest rook/rook:master
```
Comment (bassam) [inline tests/scripts/kubeadm-dind.sh]: saveimage.sh or its replacement should tag images with something stable, and you can avoid passing BUILD_NUM.

File: tests/scripts/saveImages.sh [hunk 12]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash -e
+
+scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+source "${scriptdir}/../../build/common.sh"
+
+docker save -o rookamd64.tar  ${BUILD_REGISTRY}/rook-amd64:latest
```
Comment (bassam) [inline tests/scripts/saveImages.sh]: change this to `docker tag ${BUILD_REGISTRY}/rook-amd64:latest testing/rook-amd64:latest` and the save.

also add load support in this script.

File: Jenkinsfile (hunk 5 above)
Comment (bassam) [inline Jenkinsfile]: see bottom of https://stackoverflow.com/questions/37920722/matrix-configuration-with-jenkins-pipelines for an example.

File: Jenkinsfile (hunk 1 above)
Comment (dangula) [inline Jenkinsfile]: I was thinking we would run the smoke test first, and then run a full regression on the whole matrix.

File: Jenkinsfile (hunk 2 above)
Comment (dangula) [inline Jenkinsfile]: I will keep working on this, so far Jenkins doesn't like mixing parallel and non-parallel steps, but i am looking at ways to combine the two steps.

File: tests/baseSetup.go (hunk 10 above)
Comment (dangula) [inline tests/baseSetup.go]: I am going to clean this up as part of #774  the flags names have to be consistent to use and Flasg.setwithEnv and clean up unused flags.

**Comment by dangula** [comment]
fix #786

File: tests/scripts/makeTestImages.sh [hunk 13]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash -e
+
+scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+source "${scriptdir}/../../build/common.sh"
+
+
+case "${1:-}" in
+  save)
+    case "${2:-}" in
+        arm)
```
Comment (bassam) [inline tests/scripts/makeTestImages.sh]: please simplify this. 
```
case "${2:-}" in
    arm|arm64|amd64)
      docker tag ${BUILD_REGISTRY}/rook-${2}:latest rook/rook:master
etc.
```

File: tests/scripts/makeTestImages.sh [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+            docker tag ${BUILD_REGISTRY}/rook-arm:latest rook/rook:master
+            docker tag ${BUILD_REGISTRY}/toolbox-arm:latest rook/toolbox:master
+            docker save -o rookarm.tar rook/rook:master rook/toolbox:master
```
Comment (bassam) [inline tests/scripts/makeTestImages.sh]: also preferably keep the hyphen in the tar file name rook-arm.tar

File: tests/scripts/makeTestImages.sh [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+            ;;
+        *)
+            echo "usage :" >&2
+            echo "$0 $1 arm" >&2
+            echo "$0 $1 arm64" >&2
+            echo "$0 $1 arm64" >&2
+    esac
+    ;;
+  load)
+    case "${2:-}" in
```
Comment (bassam) [inline tests/scripts/makeTestImages.sh]: simplify here too.

File: tests/scripts/makeTestImages.sh [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        *)
+            echo "usage :" >&2
+            echo "$0 $1 arm" >&2
+            echo "$0 $1 arm64" >&2
+            echo "$0 $1 arm64" >&2
+    esac
+    ;;
+  *)
+    echo "usage:" >&2
+    echo "  $0 save arm/arm64/amd64" >&2
```
Comment (bassam) [inline tests/scripts/makeTestImages.sh]: ```$0 save [arm|arm64|amd64]```

File: tests/scripts/kubeadm-dind.sh [hunk 17]
```diff
@@ -40,8 +39,9 @@ export CNI_PLUGIN="${CNI_PLUGIN:-bridge}"
 case "${1:-}" in
   up)
     ${scriptdir}/dind-cluster.sh reup
-    copy_image_to_cluster ${BUILD_REGISTRY}/rook-amd64 rook/rook:master
-    copy_image_to_cluster ${BUILD_REGISTRY}/toolbox-amd64 rook/toolbox:master
+    ${scriptdir}/makeTestImages.sh save amd64 || true
```
Comment (bassam) [inline tests/scripts/kubeadm-dind.sh]: why the ```|| true```?

File: tests/scripts/kubeadm-dind.sh [hunk 18]
```diff
@@ -7,23 +7,22 @@ tarname=image.tar
...
     done
 }
 
 copy_rbd() {
     for c in kube-master kube-node-1 kube-node-2; do
         docker cp ${scriptdir}/dind-cluster-rbd ${c}:/bin/rbd
         docker exec ${c} /bin/bash -c "chmod +x /bin/rbd"
-        docker exec ${c} /bin/bash -c "docker pull ceph/base"
+        # hack for Azure, after vm is started first docker pull command fails intermittently
+        docker exec ${c} /bin/bash -c "docker pull ceph/base || true"
```
Comment (bassam) [inline tests/scripts/kubeadm-dind.sh]: maybe use a retry instead.

File: Jenkinsfile [hunk 19]
```diff
@@ -61,3 +72,35 @@ pipeline {
...
 }
+def RunIntegrationTest(k, v) {
+  return {
+      node("${k}") {
+        script{
+            try{
+                withEnv(["KUBE_VERSION=${v}"]){
+                    unstash 'repo'
+                    echo "running tests on k8s version ${v}"
+                    sh 'tests/scripts/makeTestImages.sh  load amd64'
```
Comment (bassam) [inline Jenkinsfile]: nit: extra space.

File: Jenkinsfile [hunk 20]
```diff
@@ -61,3 +72,35 @@ pipeline {
...
+                    sh "tests/scripts/kubeadm-dind.sh up"
+                    sh 'build/run make -j\$(nproc) test-integration'
+                }
+            }
+            finally{
+                archive '_output/tests/*'
+                junit allowEmptyResults: true, keepLongStdio: true, testResults: '_output/tests/*.xml'
+                sh 'make -j\$(nproc) clean'
+                sh 'make -j\$(nproc) prune PRUNE_HOURS=48 PRUNE_KEEP=48'
+                sh 'make -j\$(nproc) -C build/release clean'
```
Comment (bassam) [inline Jenkinsfile]: you don't need `build/release clean`

File: tests/scripts/kubeadm-dind.sh (hunk 18 above)
Comment (bassam) [inline tests/scripts/kubeadm-dind.sh]: the script is running with `-e` mode. if docker exec fails will the whole script to fail? do you need to wrap it with `set +e`

Comment (dangula) [inline tests/scripts/kubeadm-dind.sh]: thanks for catching that, fixed it.
